wdl-engine/src/http.rs

```diff
@@ -300,59 +321,111 @@ impl Downloader for HttpDownloader {
-                                Status::Downloading(Arc::default()),
+                                Status::Downloading(Arc::new(Notify::new())),
                             );
-                            break;
+                            Either::Right(())
                         }
                     }
                 };
+                // Scope to ensure the mutex guard is not visible to the await point
+                let notify_to_await = match status_check_result {
+                    Either::Left(notify) => Some(notify),
+                    Either::Right(()) => None,
+                };
+
+                if let Some(notify) = notify_to_await {
+                    notify.notified().await;
+                    retried_wait = true;
+                    continue;
+                } else {
+                    break;
+                }
+            }
+
+            let mut last_error: Option<Arc<Error>> = None;
+            let mut result: Result<Location<'static>, Arc<Error>> =
+                Err(Arc::new(anyhow!("download never attempted")));
+
+            for attempt in 0..MAX_DOWNLOAD_ATTEMPTS {
+                let permit = self
+                    .semaphore
+                    .acquire()
+                    .await
+                    .expect("semaphore should not be closed");
+
+                match self.get(&url).await {
+                    Ok(location) => {
+                        result = Ok(location);
+                        break; // permit is dropped here
+                    }
+                    Err(e) => {
+                        last_error = Some(Arc::new(e));
+                        // if it was the last attempt, return the error
+                        if attempt == MAX_DOWNLOAD_ATTEMPTS - 1 {
+                            result = Err(last_error.clone().unwrap_or_else(|| {
+                                Arc::new(anyhow!("unknown download error after max attempts"))
+                            }));
+                            break;
+                        }
+
+                        // backoff and retry
+                        let delay_secs =
+                            INITIAL_RETRY_DELAY.as_secs_f64() * 2.0f64.powi(attempt as i32);
+                        let delay = Duration::from_secs_f64(delay_secs);
+                        info!(
+                            "backing off for {:.2}s before retry attempt {}/{} for `{url}`",
+                            delay_secs,
+                            attempt + 1,
+                            MAX_DOWNLOAD_ATTEMPTS,
+                            url = url
+                        );
+
+                        drop(permit);
+                        sleep(delay).await;
+                    }
+                }
+                // permit is implicitly dropped here
+            }
 
-                notify.notified().await;
-                retried = true;
-                continue;
+            if last_error.is_some() && result.is_ok() {
```

Comment:
Won't this capture the case where we errored once or twice, but then on the last try the download succeeded? In that case, do we really want to return an error?

Reply:
ah my bad i was being conservative here - removed thanks!